File: batch-worker-sdk-typescript/src/handler.ts
/**
 * Handler SPI for tenant task code (wire-protocol §4 task loop, §B field names).
 *
 * A tenant implements `TaskHandler.execute(ctx)` and returns a `TaskResult`.
 * The runtime supplies a `TaskContext` carrying the effective config, traceId,
 * a `CancellationSignal` (flipped by the lease-renewal scheduler when the
 * platform sets `cancelRequested`), and a `ProgressReporter`.
 */

import { ErrorCode } from "./protocol.ts";

/**
 * Cooperative cancellation token. The lease-renewal scheduler calls
 * `markCancelled()` when a renew response carries `cancelRequested=true`; the
 * handler polls `isCancellationRequested` at safe checkpoints and bails out.
 */
export interface CancellationSignal {
  readonly isCancellationRequested: boolean;
  markCancelled(): void;
  /** Register a callback fired once when cancellation is first requested. */
  onCancel(cb: () => void): void;
}

/** Default in-memory CancellationSignal. */
export class SimpleCancellationSignal implements CancellationSignal {
  #cancelled = false;
  #callbacks: Array<() => void> = [];

  get isCancellationRequested(): boolean {
    return this.#cancelled;
  }

  markCancelled(): void {
    if (this.#cancelled) return;
    this.#cancelled = true;
    for (const cb of this.#callbacks) {
      try {
        cb();
      } catch {
        // a stray cancel callback must never break the scheduler
      }
    }
    this.#callbacks = [];
  }

  onCancel(cb: () => void): void {
    if (this.#cancelled) {
      cb();
      return;
    }
    this.#callbacks.push(cb);
  }
}

/** Progress sink. Real impl forwards to logs / metrics; default is a no-op. */
export interface ProgressReporter {
  /** @param percent 0..100 @param message optional human note */
  report(percent: number, message?: string): void;
}

/** No-op progress reporter (records nothing). */
export class NoopProgressReporter implements ProgressReporter {
  report(_percent: number, _message?: string): void {
    // intentionally empty
  }
}

/** Execution context handed to a handler for one task. */
export interface TaskContext {
  readonly taskId: string;
  /** EffectiveTaskConfig snapshot returned by claim. */
  readonly effectiveConfig: Record<string, unknown>;
  /** runtimeAttributes.traceId for OTel correlation (may be empty). */
  readonly traceId: string;
  readonly cancellation: CancellationSignal;
  readonly progress: ProgressReporter;
}

/**
 * Task result. **Field names are a red line (wire-protocol §B):** the platform
 * `TaskExecutionReportDto` reads exactly `errorCode` / `outputs` /
 * `resultSummary`; any other name is silently dropped.
 */
export interface TaskResult {
  success: boolean;
  errorCode?: ErrorCode | string;
  outputs?: Record<string, unknown>;
  resultSummary?: string;
}

/** Build a success result. */
export function taskSuccess(
  outputs?: Record<string, unknown>,
  resultSummary?: string,
): TaskResult {
  const r: TaskResult = { success: true };
  if (outputs !== undefined) r.outputs = outputs;
  if (resultSummary !== undefined) r.resultSummary = resultSummary;
  return r;
}

/** Build a failure result. */
export function taskFailure(
  errorCode: ErrorCode | string = ErrorCode.EXECUTION_FAILED,
  resultSummary?: string,
  outputs?: Record<string, unknown>,
): TaskResult {
  const r: TaskResult = { success: false, errorCode };
  if (resultSummary !== undefined) r.resultSummary = resultSummary;
  if (outputs !== undefined) r.outputs = outputs;
  return r;
}

/** Tenant task SPI. One handler instance may serve many concurrent tasks. */
export interface TaskHandler {
  execute(ctx: TaskContext): Promise<TaskResult>;
}

File: sdk/go/client/handler.go
package client

import (
	"context"
	"sync"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/protocol"
)

// Handler SPI — the tenant-implemented surface. Mirrors the Java SDK
// TaskContext / CancellationSignal / ProgressReporter / TaskResult / TaskHandler
// (byo-sdk-guide §1.6, wire-protocol §4). Pure value + context types; no IO.

// CancellationSignal is the cooperative-cancellation primitive handed to a
// handler. Backed by a context.Context whose cancel func flips on cancel.
// IsCancellationRequested is the cheap hot-loop poll; MarkCancelled is invoked
// by the lease-renewal scheduler when renew returns cancelRequested=true.
type CancellationSignal struct {
	ctx    context.Context
	cancel context.CancelFunc
	once   sync.Once
}

// NewCancellationSignal builds a signal whose parent governs the overall task
// lifetime (e.g. worker shutdown also cancels in-flight tasks).
func NewCancellationSignal(parent context.Context) *CancellationSignal {
	if parent == nil {
		parent = context.Background()
	}
	ctx, cancel := context.WithCancel(parent)
	return &CancellationSignal{ctx: ctx, cancel: cancel}
}

// IsCancellationRequested reports whether cancellation has been signalled
// (either via MarkCancelled or the parent context being cancelled).
func (c *CancellationSignal) IsCancellationRequested() bool {
	select {
	case <-c.ctx.Done():
		return true
	default:
		return false
	}
}

// MarkCancelled flips the signal. Idempotent.
func (c *CancellationSignal) MarkCancelled() {
	c.once.Do(c.cancel)
}

// Context exposes the underlying context for select-based waits.
func (c *CancellationSignal) Context() context.Context { return c.ctx }

// ProgressReporter receives coarse-grained progress from a running handler.
// The default no-op records nothing; tests/observability can substitute one.
type ProgressReporter interface {
	// Report posts a 0..100 percentage with an optional human message.
	Report(percent int, message string)
}

// NoopProgressReporter discards progress.
type NoopProgressReporter struct{}

func (NoopProgressReporter) Report(int, string) {}

// RecordingProgressReporter captures progress events for tests.
type RecordingProgressReporter struct {
	mu     sync.Mutex
	Events []ProgressEvent
}

// ProgressEvent is one captured progress callback.
type ProgressEvent struct {
	Percent int
	Message string
}

// Report appends an event.
func (r *RecordingProgressReporter) Report(percent int, message string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.Events = append(r.Events, ProgressEvent{Percent: percent, Message: message})
}

// TaskContext is the per-task execution context passed to a handler. Carries the
// claimed EffectiveConfig snapshot, trace id for OTel stitching, the cancellation
// signal, and the progress sink.
type TaskContext struct {
	TaskID          string
	EffectiveConfig map[string]any
	TraceID         string
	Cancellation    *CancellationSignal
	Progress        ProgressReporter
}

// TaskResult is the handler's terminal outcome. Field names + json tags are a
// HARD contract (wire-protocol §B field-name 红线): the platform's
// TaskExecutionReportDto reads exactly errorCode / outputs / resultSummary;
// any other name is silently dropped.
type TaskResult struct {
	ErrorCode     protocol.ErrorCode `json:"errorCode"`
	Outputs       map[string]any     `json:"outputs"`
	ResultSummary string             `json:"resultSummary"`
}

// Success builds a SUCCESS result.
func Success(outputs map[string]any, summary string) TaskResult {
	return TaskResult{ErrorCode: protocol.ErrorCodeSuccess, Outputs: outputs, ResultSummary: summary}
}

// Fail builds a failed result with the given code + summary.
func Fail(code protocol.ErrorCode, summary string) TaskResult {
	return TaskResult{ErrorCode: code, ResultSummary: summary}
}

// IsSuccess reports whether the result is a success outcome.
func (r TaskResult) IsSuccess() bool {
	return r.ErrorCode == "" || r.ErrorCode == protocol.ErrorCodeSuccess
}

// TaskHandler is the tenant business-logic SPI. Execute runs the task and
// returns its terminal result. Implementations MUST poll
// ctx.Cancellation.IsCancellationRequested() cooperatively.
type TaskHandler interface {
	Execute(ctx *TaskContext) TaskResult
}

// HandlerFunc adapts a plain func to a TaskHandler.
type HandlerFunc func(ctx *TaskContext) TaskResult

// Execute implements TaskHandler.
func (f HandlerFunc) Execute(ctx *TaskContext) TaskResult { return f(ctx) }
